### ISSUE TITLE
k8s kill pods/containers due to protection machanism

### ISSUE THREAD

**Issue by DongZhaoYu** [Microsoft/pai#915]
K8s can kill pods or containers on a machine to protect its health.
This could be triggered by the shortage of disk space.
Need to configure the hdfs to avoid consume too much disk to trigger k8s's protection mechanism.

**Comment by DongZhaoYu** [comment]
It seems the disk pressure is not  caused by the hdfs. The pulled image seems consumed much disk storage. But need further investigation.

**Comment by fanyangCS** [comment]
Is there a way to limit docker cache?

**Comment by DongZhaoYu** [comment]
Is k8s aware of the docker container we started from Yarn container? If not the k8s GC may cannot clean the cache of the docker container.

**Comment by fanyangCS** [comment]
@DongZhaoYu  No, k8s is not aware of those docker containers.

**Comment by fanyangCS** [comment]
https://k8smeetup.github.io/docs/tasks/configure-pod-container/quality-service-pod/.
For k8s Pods, we can assign a proper QoS level to reduce the chance of critical pods being killed.

**Comment by fanyangCS** [comment]
https://github.com/Microsoft/pai/issues/827

**Comment by hao1939** [comment]
Here's stackoverflow session about limiting storage usage of a container: https://stackoverflow.com/a/33017556/500268.

Recently, docker support limit the storage on some kind of storage drivers.

https://docs.docker.com/engine/reference/commandline/run/#set-storage-driver-options-per-container
```
Set storage driver options per container
$ docker run -it --storage-opt size=120G fedora /bin/bash
This (size) will allow to set the container rootfs size to 120G at creation time. This option is only available for the devicemapper, btrfs, overlay2, windowsfilter and zfs graph drivers. For the devicemapper, btrfs, windowsfilter and zfs graph drivers, user cannot pass a size less than the Default BaseFS Size. For the overlay2 storage driver, the size option is only available if the backing fs is xfs and mounted with the pquota mount option. Under these conditions, user can pass any size less than the backing fs size.
```

**Comment by hao1939** [comment]
#1777 is for the disk restriction.

**Comment by fanyangCS** [comment]
fix in 0.10.y by: 1 disable pod eviction (both kubelet and cluster level). 2 deploy a watchdog to kill job containers when it consumes too much resource. 3. enforce max memory usage for each PAI service pod.
